feat(Computability/NFA): nfa closure properties

rudynicolop opened the pull request:
Add the closure properties union, intersection and reversal for NFA.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

rudynicolop committed:
NFA closure properties.

rudynicolop commented:
As per the discussion on #22298, this only includes changes to NFA.

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -85,12 +85,24 @@ theorem eval_singleton (a : α) : M.eval [a] = M.stepSet M.start a :=
 theorem eval_append_singleton (x : List α) (a : α) : M.eval (x ++ [a]) = M.stepSet (M.eval x) a :=
   evalFrom_append_singleton _ _ _ _
 
+/--
+`M.acceptsFrom S` is the language of `x` such that `M.evalFrom S x` is an accept state.
+-/
```
This is more common (and sticks to the style of the file).
```suggestion
/-- `M.acceptsFrom S` is the language of `x` such that `M.evalFrom S x` is an accept state. -/
```

rudynicolop replied:
Ok thanks, I think I fixed it.

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -85,12 +85,24 @@ theorem eval_singleton (a : α) : M.eval [a] = M.stepSet M.start a :=
   evalFrom_append_singleton _ _ _ _
 
+/--
+`M.acceptsFrom S` is the language of `x` such that `M.evalFrom S x` is an accept state.
```
Rather this?
```suggestion
`M.acceptsFrom S` is the language of `x` such that `M.evalFrom S x` contains an accept state.
```

rudynicolop replied:
Yeah, thanks!.

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -85,12 +85,24 @@ theorem eval_singleton (a : α) : M.eval [a] = M.stepSet M.start a :=
+-/
+def acceptsFrom (S : Set σ) : Language α := {x | ∃ s ∈ M.accept, s ∈ M.evalFrom S x}
+
+theorem mem_acceptsFrom {S : Set σ} {x : List α} :
+  x ∈ M.acceptsFrom S ↔ ∃ s ∈ M.accept, s ∈ M.evalFrom S x := by rfl
```
The statement of a theorem has to be indented with 4 spaces.
```suggestion
theorem mem_acceptsFrom {S : Set σ} {x : List α} :
    x ∈ M.acceptsFrom S ↔ ∃ s ∈ M.accept, s ∈ M.evalFrom S x := by rfl
```

rudynicolop replied:
Ok, sounds good! I think it fixed it throughout the file.

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -85,12 +85,24 @@ theorem eval_singleton (a : α) : M.eval [a] = M.stepSet M.start a :=
 theorem mem_accepts {x : List α} : x ∈ M.accepts ↔ ∃ S ∈ M.accept, S ∈ M.evalFrom M.start x := by
   rfl
 
+lemma accepts_acceptsFrom :
+  M.accepts = M.acceptsFrom M.start := by
+  dsimp [accepts, acceptsFrom, eval]
```
simpler:
```suggestion
lemma accepts_acceptsFrom : M.accepts = M.acceptsFrom M.start := rfl
```

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+section Auxilary
+
+/-- `M unstep` is the reverse of `M.step`
+-/
```
Don't skip lines for delimiters of docstring. Also add the full stop please. And it is `M.unstep` rather than `M unstep`.
```suggestion
/-- `M.unstep` is the reverse of `M.step`. -/
```

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+/-- `M unstep` is the reverse of `M.step`
+-/
+def unstep (s : σ) (a : α) : Set σ :=
+  fun s' ↦ M.step s' a s
```
We avoid spelling sets as functions. Also you can put everything on one line.
```suggestion
def unstep (s : σ) (a : α) : Set σ := {s' | s ∈ M.step s' a}
```

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+def unstep (s : σ) (a : α) : Set σ :=
+  fun s' ↦ M.step s' a s
+
+theorem mem_unstep (s t : σ) (a : α) :
+  s ∈ M.unstep t a ↔ t ∈ M.step s a := by
+  rfl
```
Again you can put everything on one line. We tend to minimize explicit argument in `iff` results to simplify the use of dot notation, here all your variables appear on both sides of the `iff` so you can make them implicit. Also note that all the proofs which go as `by rfl` can be replaced by `Iff.rfl`, I would say the latter is more common in the library but you can choose the version you prefer.
```suggestion
theorem mem_unstep {s t : σ} {a : α} : s ∈ M.unstep t a ↔ t ∈ M.step s a := Iff.rfl
```

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  s ∈ M.unstep t a ↔ t ∈ M.step s a := by
+  rfl
+
+/-- `M.unstepSet S a` is the union of `unstep M s a` for all `s ∈ S`. -/
```
Can you write a more enlightening docstring please? Here you just describe the equation below, I think it would be more helpful if you explained what it is in terms of states and paths and all.

rudynicolop replied:
Ok, I have a new docstring.

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+def unstepSet (S : Set σ) (a : α) : Set σ :=
+  ⋃ s ∈ S, M.unstep s a
+
+theorem mem_unstepSet (s : σ) (S : Set σ) (a : α) :
+  s ∈ M.unstepSet S a ↔ ∃ t ∈ S, s ∈ M.unstep t a := by
+  simp [unstepSet]
```
Again I would make everything explicit here, and mind the indentation. This comment apply to many results so can you please go carefully through all of them and apply it when needed?
```suggestion
theorem mem_unstepSet {s : σ} {S : Set σ} {a : α} :
    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, s ∈ M.unstep t a := by
  simp [unstepSet]
```

rudynicolop replied:
> I would make everything explicit here

Double-checking: do you mean that the parameters to theorems and lemmas should generally *implicit*?
Looking over the file at previous theorems and lemmas, do we generally want dependent parameters to be *implicit* unless the lemma or theorem is marked with `@[simp]`?

Thanks!

EtienneC30 replied:
My bad I meant *implicit* here. The rules for impliciteness are not always clear (for me at least), but in this particular case here is how it goes: in an `iff` lemma you want to avoid as much explicit parameters as possible, to be able to write `mem_unstepSet.mp` when referring to the forward direction and not `(mem_unstepSet s S a).mp`. Of course the issue with implicit parameters is that if they are not inferred from the context then they have to be given manually which can make code heavier. In this particular case however, all the parameters appear on both sides of the `iff` and so you can be pretty sure that they will be inferred from the context, so you can make them implicit.

This is not related to `@[simp]`.

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+theorem mem_unstepSet_step (s : σ) (S : Set σ) (a : α) :
+  s ∈ M.unstepSet S a ↔ ∃ t ∈ S, t ∈ M.step s a := by
+  rw [mem_unstepSet]
+  apply exists_congr
+  intro t
+  rw [mem_unstep]
```
Shorter proof.
```suggestion
  simp [mem_unstepSet, mem_unstep]
```

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  intro t
+  rw [mem_unstep]
+
+/-- `M.rewindFrom S x` computes all possible paths though `M` with input `x` ending at an element
+  of `S`. -/
+def rewindFrom (final : Set σ) : List α → Set σ :=
```
Typo. Also can you use the same variable in the docstring and in the def please (I'm talking about `S` vs `final`)? I'm not sure the docstring really describes the function. For me it rather computes a set of state rather than a set of paths.
```suggestion
/-- `M.rewindFrom S x` computes all possible paths through `M` with input `x` ending at an element
  of `S`. -/
def rewindFrom (final : Set σ) : List α → Set σ :=
```

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  of `S`. -/
+def rewindFrom (final : Set σ) : List α → Set σ :=
+  List.foldl M.unstepSet final
+
```
```suggestion
```

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  List.foldl M.unstepSet final
+
+
+/-- `M.rewind x` computes all possible paths though `M` with input `x` ending at an element of
+  `M.accept`. -/
```
```suggestion
/-- `M.rewind x` computes all possible paths through `M` with input `x` ending at an element of
  `M.accept`. -/
```

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+/-- `startsTo M S` is the language of `x`
+such that starting from `S` we arrive at `M.start`.
+-/
+def startsTo (S : Set σ) : Language α :=
```
I'm not sure about the name, does "starting to" means something in English? Maybe `startsFrom` is better.

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+def rewind : List α → Set σ :=
+  M.rewindFrom M.accept
+
+/-- `startsTo M S` is the language of `x`
```
```suggestion
/-- `M.startsTo S` is the language of `x`
```

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+/-- NFAs are closed under reversal:
+Given NFA `M`, there is an NFA `reverse(M)` such that `L(reverse(M)) = reverse(L(M))`.
+-/
+def reverse (M : NFA α σ) : NFA α σ :=
+  NFA.mk M.unstep M.accept M.start
```
Can you use the `where` syntax please?

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+def reverse (M : NFA α σ) : NFA α σ :=
+  NFA.mk M.unstep M.accept M.start
+
+lemma reverse.SpecFrom (S : Set σ) (M : NFA α σ) :
```
Namespaces have to start with a capital letter. Besides, if we do this we can't use dot notation anymore, so I'd suggest to remove the `reverse.`. Also lemmas are written in `lower_camel_case`.
```suggestion
lemma spec_from (S : Set σ) (M : NFA α σ) :
```

rudynicolop replied:
Ok, I think I removed the namespaces I added.

> Also lemmas are written in `lower_camel_case`.

Double-checking: do you mean `lower_snake_case` for lemmas and theorems, and `lowerCamelCase` for definitions?

Thanks!

EtienneC30 replied:
Yes my bad. The naming for definitions actually depends on the return type, see the [naming convention](https://leanprover-community.github.io/contribute/naming.html#general-conventions).

EtienneC30 commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +124,309 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  NFA.mk M.unstep M.accept M.start
+
+lemma reverse.SpecFrom (S : Set σ) (M : NFA α σ) :
+  M.reverse.acceptsFrom S =
+  { xs : List α | xs ∈ M.startsTo S } := by
```
You should rather state that (and `rfl` closes the goal):
```suggestion
  M.reverse.acceptsFrom = M.startsTo := by
```

EtienneC30 reviewed (commented):
Thanks for your work. Here comes a first wave of comments. I only looked at the beginning but many comments should apply to all the file, so make sure to go through it carefully. Don't hesitate to ask if you have any questions.

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

comment style

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

indentation style

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

simplify proof

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

corrected comment

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

comment style

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

set style

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

code style

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

indentation and implicitness

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

simplify proof

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

spelling error

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

spelling error

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

rename reverse spec

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

simplify definition

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
fixed build errors from review suggestions: todo: fix style

rudynicolop commented:
Thanks for the feedback! I will begin working on it!

To summarize, I think that these are the current changes you are suggesting:

PR todos:
- [ ] The statement of a theorem has to be indented with 4 spaces.
- [ ] Simplify proofs.
- [ ] Don't skip lines for delimiters of docstring.
- [ ] We avoid spelling sets as functions.
- [ ] Put everything on one line when possible.
- [ ] Some proofs which go as by `rfl` can be replaced by `Iff.rfl`.
- [ ] Make dependent parameters of theorems and lemmas implicit.
- [ ] Rethink name for `startsTo`.
- [ ] Use `where` syntax for `reverse` def.
- [ ] Namespaces have to start with a capital letter. Besides, if we do this we can't use dot notation anymore, so I'd suggest to remove the `reverse.`. Also lemmas are written in lower_camel_case.
- [ ] Better docstring for `[unstepSet]`: explain what it is in terms of states and paths and all.
- Better docstring for `[rewindFrom]`: 
  * [ ] Use the same variable in the docstring and in the def please (I'm talking about `S` vs `final`)?
  * [ ] I'm not sure the docstring really describes the function. For me it rather computes a set of state rather than a set of paths.

rudynicolop committed:
no longer skip lines for the delimiter of a docstring

rudynicolop committed:
fix indentation in reverse docstring

rudynicolop committed:
indent lemma/theorem statement 4 spaces

rudynicolop committed:
straggler: indent reverse.Spec 4 spaces

rudynicolop committed:
Iff.rfl in place of by rfl

rudynicolop committed:
where syntax for NFA reverse, union, intersect

rudynicolop committed:
startsTo renamed to rewindsToStart

rudynicolop committed:
Update Mathlib/Computability/NFA.lean

remove extra whitespace

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

rudynicolop committed:
better docstring for rewindsFrom

rudynicolop committed:
better docstring for unstepSet

rudynicolop committed:
better docstring for unstepSet

rudynicolop committed:
remove extra namespaces and lower snake case for theorem and lemmas

rudynicolop committed:
intersect lemma name style consistent with union

rudynicolop committed:
made explicit dependent parameters of theorems and lemmas implicit

rudynicolop committed:
formatting, simplified some proofs

rudynicolop committed:
tried to use set membership instead of functions, some internal lemmas still use function application

rudynicolop commented:
Thanks for the feedback @EtienneC30!

I believe I addressed your suggestions. Apologies if I missed anything.

Should I add myself as an `Author` at the top?

Thanks again!

rudynicolop committed:
author

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +120,275 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  ext xs
+  rfl
+
+lemma reverse_rewindFrom_evalFrom {xs : List α} {S1 S2 : Set σ} {M : NFA α σ} :
+    (∃ s1 ∈ S1, s1 ∈ M.rewindFrom S2 xs) ↔
+    (∃ s2 ∈ S2, s2 ∈ M.evalFrom S1 xs.reverse) := by
+  dsimp [evalFrom, rewindFrom]
+  rw [List.foldl_reverse]
+  revert S1 S2
+  induction xs <;> intros S1 S2 <;> simp
+  case nil =>
+    constructor <;> rintro ⟨s, h1, h2⟩ <;> exists s
+  case cons x xs ih =>
+    rw [ih]
+    clear ih
+    unfold unstepSet
+    unfold stepSet
+    constructor
+    · rintro ⟨s3, h, h1⟩
+      simp [Set.mem_iUnion] at h
+      rcases h with ⟨s2,h2,h3⟩
+      exists s2
+      constructor <;> try assumption
+      simp [Set.mem_iUnion]
+      exists s3
+    · rintro ⟨s2, h2, h⟩
+      simp [Set.mem_iUnion] at h
+      rcases h with ⟨s3, h1, h3⟩
+      exists s3
+      simp [Set.mem_iUnion]
+      constructor <;> try assumption
+      exists s2
```
Golfed:

```lean
lemma reverse_rewindFrom_evalFrom {xs : List α} {S1 S2 : Set σ} {M : NFA α σ} :
    (∃ s1 ∈ S1, s1 ∈ M.rewindFrom S2 xs) ↔
    (∃ s2 ∈ S2, s2 ∈ M.evalFrom S1 xs.reverse) := by
  dsimp [evalFrom, rewindFrom]
  rw [List.foldl_reverse]
  induction xs generalizing S1 S2
  case nil => tauto
  case cons x xs ih =>
    simp [ih, unstepSet, stepSet, Set.mem_iUnion]
    tauto
```

Of note:
- `revert; induction; intros` is just `induction ... generalizing`
- I can't find it documented right now, but I believe it is not good practice to have [unsqueezed `simp` calls in non-terminal positions](https://leanprover-community.github.io/contribute/style.html#squeezing-simp-calls) because such proofs break easily.

meithecatte replied:
Also, the statement is quite weird. "exists s1 in S1 such that s1 is also in blah" is simply "S1 is a subset of blah". I think that's how you want to do this? Or maybe prove a more general `S₁ ⊆ S₂ -> M.evalFrom S₁ x ⊆ M.evalFrom S₂ x` and use that to simplify things.

rudynicolop replied:
Thanks for the feedback! I will try the subset style definition.

rudynicolop replied:
>I can't find it documented right now, but I believe it is not good practice to have [unsqueezed simp calls in non-terminal positions](https://leanprover-community.github.io/contribute/style.html#squeezing-simp-calls) because such proofs break easily.

What is the general fix for this?

meithecatte replied:
> What is the general fix for this?

I found [some more relevant documentation](https://leanprover-community.github.io/extras/simp.html#non-terminal-simps) after I wrote that comment, perhaps that'll be helpful.

meithecatte replied:
> Thanks for the feedback! I will try the subset style definition.

oh, sorry, looks like I've misled you, I had a brainfart. what I'm saying would be true if it was a forall. since it's an exists, you could state it as "the intersection of these two sets is nonempty". not sure which would work better, but writing this as an intersection seems like it would be more flexible (because intersection is easily commutative). your call.

rudynicolop replied:
> > Thanks for the feedback! I will try the subset style definition.
> 
> oh, sorry, looks like I've misled you, I had a brainfart. what I'm saying would be true if it was a forall. since it's an exists, you could state it as "the intersection of these two sets is nonempty". not sure which would work better, but writing this as an intersection seems like it would be more flexible (because intersection is easily commutative). your call.

What do you mean? Is it ok to leave as is?

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +120,275 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
 
+section Closure
+
+/-! ### NFA Closure Properties -/
```
It would be good to add a short doc comment along the lines of:

> Properties of the form "if these languages can be represented by NFAs, then these related languages also can".

This is to avoid confusion with e.g. Kleene's closure.

rudynicolop replied:
Ok, sounds good. I added something but I am not sure of the best way to phrase it nor if the documentation formatting/style is good.

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +120,275 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    (⋃ s, ⋃ (_ : Sum.rec S1 S2 s), stepSum M1 M2 s x)
+    =
+    {s' | Sum.rec
+      (fun s1' ↦ (⋃ (s1 : σ1) (_ : s1 ∈ S1), M1.step s1 x) s1')
+      (fun s2' ↦ (⋃ (s2 : σ2) (_ : s2 ∈ S2), M2.step s2 x) s2') s'} := by
```
You're still treating sets as functions here.

rudynicolop replied:
Ah, sorry, thanks for catching this! I changed the lemma statement to what you suggested in the other comment.

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +120,275 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+      rcases hs2 with ⟨hs2,h2⟩
+      exists s2
+
+lemma union_start_spec {M1 : NFA α σ1} {M2 : NFA α σ2} :
```
This should use subscripts: `σ₁`, `σ₂`

rudynicolop replied:
In general are subscripts preferred?

rudynicolop replied:
Should `S1`, `S2`, `M1`, `M2`, etc also have subscripts?

meithecatte replied:
AFAIK subscripts are in general preferred, but FWIW this is something I've picked up by looking at code and couldn't find any documentation to confirm this.

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +120,275 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    simp
+    rw [←Set.mem_def (a := s1') (s := (⋃ (s1 : σ1) (_ : s1 ∈ S1), M1.step s1 x))]
+    rw [Set.mem_iUnion]
+    constructor
+    · rintro ⟨s1, hs1, h1⟩
+      exists s1
+      rw [Set.mem_iUnion]
+      exists hs1
+    · rintro ⟨s1, hs1⟩
+      rw [Set.mem_iUnion] at hs1
+      rcases hs1 with ⟨hs1,h1⟩
+      exists s1
```
This proof can also be shortened a lot by using `tauto`, like above. For example, with a statement that is somewhat but not fully exorcised from the "sets are just functions into prop" abstraction violation:

```lean
lemma union_biUnion_spec
  {x : α} {S1 : Set σ1} {S2 : Set σ2} {M1 : NFA α σ1} {M2 : NFA α σ2} :
    (⋃ s, ⋃ (_ : Sum.rec S1 S2 s), stepSum M1 M2 s x)
    =
    {s' | Sum.rec
      (fun s1' ↦ s1' ∈ ⋃ (s1 ∈ S1), M1.step s1 x)
      (fun s2' ↦ s2' ∈ ⋃ (s2 ∈ S2), M2.step s2 x) s'} := by
  ext s'
  rw [Set.mem_iUnion, Set.mem_setOf]
  dsimp [stepSum]
  cases s' <;> simp <;> tauto
```

rudynicolop replied:
Thanks! I used your version of the lemma statement: maybe it's possible to further exorcise "sets are just functions into prop" but maybe this is acceptable?

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -112,6 +120,275 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+lemma union_start_spec {M1 : NFA α σ1} {M2 : NFA α σ2} :
+    (union M1 M2).start = { s : σ1 ⊕ σ2 | s.casesOn M1.start M2.start } := by rfl
+
+lemma union_acceptsFrom
+ {S1 : Set σ1} {S2 : Set σ2} {M1 : NFA α σ1} {M2 : NFA α σ2} :
+    acceptsFrom (union M1 M2)
+      { s : σ1 ⊕ σ2 | s.casesOn S1 S2 }
+    = M1.acceptsFrom S1 ∪ M2.acceptsFrom S2 := by
+    apply Set.ext
+    intros xs
+    dsimp [acceptsFrom, evalFrom, union, accept]
+    unfold stepSet
+    dsimp [NFA.step]
+    rw [Set.mem_union, Set.mem_setOf, Set.mem_setOf]
+    revert S1 S2
+    induction xs <;> intro S1 S2
+    case nil =>
+      simp
+      simp [Set.mem_def]
+    case cons x xs ih =>
+      simp [List.foldl_cons, List.foldl_cons, List.foldl_cons, ←ih, union_biUnion_spec]
```
Please observe the indentation [described in the style guide](https://leanprover-community.github.io/contribute/style.html#tactic-mode)

rudynicolop replied:
My apologies, thanks for catching this!

rudynicolop committed:
simplified reverse_rewindFrom proof

rudynicolop committed:
fixed indentation

rudynicolop committed:
updated nfa closure doc string

rudynicolop committed:
functions as sets changed to set membership, simplified proofs with tauto

rudynicolop commented:
Apologies for taking a while to address the feedback, thank you so much for your help and patience! I believe I have fixed the issues caught by @meithecatte.

rudynicolop committed:
fixed docstring formatting

meithecatte commented:
GitHub for some reason believes that this PR has 1600 commits and changes 4300 files, I'll try closing the PR and immediately reopening it :shrug:

rudynicolop commented:
> GitHub for some reason believes that this PR has 1600 commits and changes 4300 files, I'll try closing the PR and immediately reopening it 🤷

My apologies, I guess I screwed something up when I rebased the changes from `master`. Thanks for the quick fix!

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +123,217 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+variable {σ1 : Type v} {σ2 : Type v}
+
+private instance Language.instUnion : Union (Language α) := by
+  apply Set.instUnion
```
Why are you putting an instance about `Language` here? Also, it's private, but you're using it to state the main theorems about what `NFA.union` does, so... how is anyone ever supposed to make use of `NFA.union`?

Also, notation for the union of languages already exists, and it's `l₁ + l₂`.

rudynicolop replied:
My apologies, I did not catch that. I replaced this instance with the `Add` instance for `Language`, and I moved the `Inter` instance I created to `Language.lean`.

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +123,217 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  Given NFAs `M₁` and `M₂`, `M₁ ∩ M₂` is a NFA such that
+  `L(M₁ ∩ M₂) = L(M₁) ∩ L(M₂)`. -/
+def intersect (M1 : NFA α σ1) (M2 : NFA α σ2) : NFA α (σ1 × σ2) where
+  start : Set (σ1 × σ2) := { s : σ1 × σ2 | s.1 ∈ M1.start ∧ s.2 ∈ M2.start }
```
```suggestion
  start : Set (σ1 × σ2) := M1.start ×ˢ M2.start
```

The same applies to everywhere else you're doing a cartesian product manually.

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +123,217 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  rw [mem_rewindsToStart, mem_acceptsFrom]
+  apply reverse_rewindFrom_evalFrom
+
+theorem reverse_accepts {M : NFA α σ} :
+    M.reverse.accepts = { xs : List α | xs.reverse ∈ M.accepts } := by
```
```suggestion
theorem reverse_accepts {M : NFA α σ} :
    M.reverse.accepts = M.accepts.reverse := by
```

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -88,12 +88,20 @@ theorem eval_singleton (a : α) : M.eval [a] = M.stepSet M.start a :=
 theorem eval_append_singleton (x : List α) (a : α) : M.eval (x ++ [a]) = M.stepSet (M.eval x) a :=
   evalFrom_append_singleton _ _ _ _
 
+/-- `M.acceptsFrom S` is the language of `x` such that `M.evalFrom S x` is an accept state. -/
```
```suggestion
/-- `M.acceptsFrom S` is the language of `x` such that there is an accept state in `M.evalFrom S x`. -/
```

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -88,12 +88,20 @@ theorem eval_singleton (a : α) : M.eval [a] = M.stepSet M.start a :=
+theorem mem_acceptsFrom {S : Set σ} {x : List α} :
+    x ∈ M.acceptsFrom S ↔ ∃ s ∈ M.accept, s ∈ M.evalFrom S x := Iff.rfl
+
 /-- `M.accepts` is the language of `x` such that there is an accept state in `M.eval x`. -/
 def accepts : Language α := {x | ∃ S ∈ M.accept, S ∈ M.eval x}
 
 theorem mem_accepts {x : List α} : x ∈ M.accepts ↔ ∃ S ∈ M.accept, S ∈ M.evalFrom M.start x := by
   rfl
 
+lemma accepts_acceptsFrom : M.accepts = M.acceptsFrom M.start := rfl
+
```
Now that `acceptsFrom` exists, `accepts` can be simply defined as `M.acceptsFrom M.start`. This is how it's done in DFA, c.f. 73756969.

rudynicolop replied:
Yes, true. Admittedly, I am not overly fond of potentially repairing other proofs in the file if I replace this definition... :grimacing: 

But maybe it's work the change I'm not sure.

meithecatte replied:
That's the thing: so far, everything uses `mem_accepts` instead of unfolding the definition, and because the equality between the old and new definitions holds by reflexivity, the set of def-eqs also doesn't change.

rudynicolop replied:
Ah, good point, my bad. I followed your suggestion.

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +123,217 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  ext xs
+  rfl
+
+lemma reverse_rewindFrom_evalFrom {xs : List α} {S1 S2 : Set σ} {M : NFA α σ} :
```
Calling this lemma `NFA.reverse_rewindFrom_evalFrom` suggests that it pertains to `NFA.reverse`, however `NFA.reverse` does not occur in the statement of the lemma. How about `rewindFrom_iff_evalFrom_reverse`?

rudynicolop replied:
Good suggestion, thanks!

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +123,217 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    simp [ih, unstepSet, stepSet, Set.mem_iUnion]
+    tauto
+
+lemma reverse_rewindsToStart_acceptsFrom {xs : List α} {M : NFA α σ} :
+    xs ∈ M.rewindsToStart M.reverse.start ↔ xs.reverse ∈ M.acceptsFrom M.start := by
+  rw [mem_rewindsToStart, mem_acceptsFrom]
+  apply reverse_rewindFrom_evalFrom
```
```suggestion
lemma mem_rewindsToStart_iff_reverse_mem_acceptsFrom {xs : List α} {M : NFA α σ} :
    xs ∈ M.rewindsToStart M.accept ↔ xs.reverse ∈ M.acceptsFrom M.start := by
  simp [mem_rewindsToStart, mem_acceptsFrom, rewindFrom_iff_evalFrom_reverse]
```

rudynicolop committed:
use language add, and moved language inter

meithecatte commented on Mathlib/Computability/Language.lean:
```diff
@@ -85,6 +85,11 @@ instance : Add (Language α) :=
 instance : Mul (Language α) :=
   ⟨image2 (· ++ ·)⟩
 
+/-- The intersection of two languages `l` and `m` is the language made of the
+strings `x` where `x ∈ l` and `x ∈ m`. -/
+instance instInter : Inter (Language α) :=
+  ⟨((· ∩ ·) : Set (List α) → Set (List α) → Set (List α))⟩
+
```
Other instances don't have names. Does this one need a name?

rudynicolop replied:
Fair point. Admittedly I personally don't like the implicit naming, but consistency is good.

meithecatte replied:
I mean, when would you need to explicitly refer to the instance, anyway?

rudynicolop replied:
In principle if I am trying to unfold a specific instance's implementation I prefer if it's not the generated name, but this probably doesn't come up too often...

rudynicolop replied:
Fixed btw.

meithecatte replied:
> In principle if I am trying to unfold a specific instance's implementation I prefer if it's not the generated name, but this probably doesn't come up too often...

Looks like the preferred solution is to have something like e.g. `Language.add_def`.

rudynicolop committed:
Apply suggestions from code review

Co-authored-by: Maja Kądziołka <github@compilercrim.es>

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +123,217 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
 
+section Closure
+
+/-! ### NFA Closure Properties:
+    Properties relating operations on NFAs to their
+    languages. -/
```
Avoid indenting comments. This file happens to already do so for the doc-comments of declarations, so those can be cleaned up later, but the top-level comment conforms to the mathlib style, so this comment should as well. (cf. https://leanprover-community.github.io/contribute/style.html#comments)

As for the wording here, I'd suggest:

```suggestion
/-!
### NFA Closure Properties

Constructions of NFAs corresponding to operations on languages.
-/
```

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +123,217 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+end Auxilary
+
+/-- NFAs are closed under reversal:
+  Given NFA `M`, there is an NFA `reverse(M)` such that
+  `L(reverse(M)) = reverse(L(M))`. -/
```
Let's use the notation that mathlib actually uses:

```suggestion
/-- NFAs are closed under reversal:
Given NFA `M`, there is an NFA `M.reverse` such that
`M.reverse.accepts = M.accepts.reverse`. -/
```

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +123,209 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, t ∈ M.step s a := by
+  simp [mem_unstepSet, mem_unstep]
+
+/-- Reseversed analog of `M.evalFrom S x`:
+  `M.rewindFrom S x` computes all possible reversed paths through `M` with
+  input `x` starting at an element of `S`. -/
```
This needs to describe from which direction you're reading the `x`, because it can be confusing. Also, you've typoed `Reversed`.

rudynicolop replied:
Yeah, good point!

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +123,209 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  input `x` starting at an element of `S`. -/
+def rewindFrom : Set σ → List α → Set σ := List.foldl M.unstepSet
+
+/-- `M.rewind x` computes all possible paths through `M` with input `x` ending at an element of
+  `M.accept`. -/
+def rewind : List α → Set σ := M.rewindFrom M.accept
```
This comment doesn't seem very accurate.

rudynicolop replied:
Ah, sorry, I confused myself :sweat_smile:. I think made it consistent with the comment for `rewindFrom`.

rudynicolop committed:
fixed build issues

rudynicolop committed:
Apply suggestions from code review

Co-authored-by: Maja Kądziołka <github@compilercrim.es>

rudynicolop committed:
fixed comments and changed lemma name

rudynicolop committed:
replaced cartesian product instances with notation

rudynicolop committed:
define NFA.accepts in terms of NFA.acceptsFrom

meithecatte commented on Mathlib/Computability/NFA.lean:
```diff
@@ -115,6 +122,210 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  simp [mem_unstepSet, mem_unstep]
+
+/-- Reversed analog of `M.evalFrom S x`:
+  `M.rewindFrom S x` computes all possible reversed paths through `M` with
+  input `x` starting at an element of `S`.
+  Even though `M.rewindFrom S x` voyages across `M` in reverse, `x` is processed
+  forward, from left to right. -/
```
I gave some thought into how to properly convey this, what do you think about this formulation?

```suggestion
  `M.rewindFrom S x` is the set of all starting points such that the automaton could be in one of the states
  in `S` after consuming `x.reverse`. -/
```

rudynicolop replied:
So this is definitely true, as proven in `rewindFrom_iff_evalFrom_reverse`, but I think that the docstring for `rewindFrom` should still say exactly what's doing. It's not directly reversing the input string `x`, but going backwards through the transitions via `unstep`.
My comment could probably be better phrased, but it should probably convey that:
- `x` is consumed in order, not-reversed, from left to right,
- It goes through the transitions of `M` backwards, starting from states in `S`

b-mehta commented:
Thanks for your PR! Computability PRs can move pretty slowly unfortunately, so thanks for your patience on this one! I've left a few comments (and it looks like there's a few comments you'd still like to address from @meithecatte), so I've marked it awaiting-author for now. After you've addressed them, feel free to remove that label and ping me for review, and let me know if you have any questions!

(Also, thanks @meithecatte for your helpful review too, I didn't look at all of your comments but the ones I did look at were good suggestions!)

meithecatte commented:
I'd also like to add that the PR is awaiting resolution of the Zulip discussion: [#mathlib4 > Regular languages: the review queue](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Regular.20languages.3A.20the.20review.20queue/with/514669532)

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

ctchou reviewed (changes requested):
For proving the closure of regular languages under union and intersection, I think a much simpler approach is to work on DFA, rather than NFA.  Both can be proved using a simple product construction on automata:
https://github.com/leanprover/cslib/pull/142/files#diff-d35d806c375ddfd6e5c40b9a8401f572c5ecdcd4bac50b772a1f33c1c1498305
Then the union and intersection of the sets of accepting states will implement the union and intersection of the accepted languages:
https://github.com/leanprover/cslib/pull/142/files#diff-c7335454dacd36c73c8a5b8c2d730270a0933d82c7c1c0b72a247e820b791fed
The whole proof can probably be done in 50 lines (if that).

Also, to be consistent with the rest Language.lean, I think you should use "add" and `+` rather than "union" and `∪` to refer to set union, and use the lattice-theoretic "inf" and `⊓` to refer to set intersection.
